### PR TITLE
More fixes and implementations for mwcc/mwld

### LINKED ISSUE #20: Support mwcceppc.exe (Metrowerks C/C++ Compiler for Embedded PowerPC)
Currently, this program fails with the following output:

```
### mwcceppc.exe Compiler Error:
#   ### Error: Cannot open main file ###

Errors caused tool to abort.
```

Test case: [mwcc_example.zip](https://github.com/user-attachments/files/16355195/mwcc_example.zip)
`retrowin32 mwcceppc.exe -nodefaults -c in.c -o out.o`
(Requires https://github.com/evmar/retrowin32/pull/19)

The included mwcceppc.exe is a version containing CodeView debugging info, so it should be easier to analyze. Here's an exported Ghidra project with many data types already set up: [mwcceppc.gzf.zip](https://github.com/user-attachments/files/16355218/mwcceppc.gzf.zip) (extract to .gzf then import)

The error is called @ 43f4a3, after checking for a pointer's validity:
```c
if (cparamblkptr->sourcetext == (char *)0x0) {
  // ...
  _CError_CannotOpen(); // 43f4a3
```
If I'm analyzing it correctly, that pointer should have been set via the call @ 41fcbb:
```c
iVar1 = _CWGetMainFileText@12
                  (param_1,&CompilerLinkerParamBlk_005e3eba.sourcetext,
                   &CompilerLinkerParamBlk_005e3eba.sourcetextsize);
```
I think something odd is happening in between.

Also of note is that the included `lmgr326b.dll` is a stub, each function exported is just `return 0;`. Hopefully isn't related, but thought it was worth mentioning.

### PR DESCRIPTION
**Pull request by encounter**
Each commit is self-contained and self-describing. (However, the last commit adding implementations and refactoring the host API was difficult to separate.) Feel free to veto any of these changes.

All together, this resolves #20.

### CODE REVIEW

File: x86/src/ops/table.rs [hunk 1]
```diff
@@ -272,6 +273,8 @@ pub unsafe fn init_op_tab() {
     OP_TAB[iced_x86::Code::Dec_r32 as usize] = Some(dec_rm32);
     OP_TAB[iced_x86::Code::Dec_rm32 as usize] = Some(dec_rm32);
     OP_TAB[iced_x86::Code::Dec_r16 as usize] = Some(dec_rm16);
+    OP_TAB[iced_x86::Code::Dec_rm16 as usize] = Some(dec_rm16);
+    OP_TAB[iced_x86::Code::Dec_r16 as usize] = Some(dec_rm8);
```
Comment (evmar) [inline x86/src/ops/table.rs:277]: dec_rm16 here

File: win32/src/winapi/kernel32/init.rs [hunk 2]
```diff
@@ -303,7 +303,7 @@ impl State {
 
     pub fn get_process_heap<'a>(&'a mut self, memory: &mut MemImpl) -> &mut Heap {
         if self.process_heap == 0 {
-            let size = 16 << 20;
+            let size = 24 << 20;
```
Comment (evmar) [inline win32/src/winapi/kernel32/init.rs:306]: My eventual idea here (not that you need to fix it, just FYI) is that we could implement heaps as multiple not necessarily contiguous regions.  Like once we fill the first Xmb we just allocate a new block of Xmb and start putting new allocations there.  Kinda gross though.

File: cli/src/host.rs [hunk 3]
```diff
@@ -20,6 +20,10 @@ impl win32::File for File {
             Err(ref e) => Err(io_error_to_win32(e)),
         }
     }
+
+    fn set_len(&self, len: u64) -> Result<(), u32> {
```
Comment (evmar) [inline cli/src/host.rs:24]: Windows trivia: extending a file's length has all these funny interactions with their async IO, see "extend a file" https://learn.microsoft.com/en-us/previous-versions/troubleshoot/windows/win32/asynchronous-disk-io-synchronous

File: win32/src/winapi/stack_args.rs [hunk 4]
```diff
@@ -49,6 +49,30 @@ impl<'a, T: FromArg<'a>> FromStack<'a> for T {
     }
 }
 
+impl<'a> FromArg<'a> for u8 {
```
Comment (evmar) [inline win32/src/winapi/stack_args.rs:52]: I wonder if we could get away with `FromArg<T> where T: Pod`.  I guess it's only Pod <32 bits though.  This is fine too.

File: cli/src/host.rs [hunk 5]
```diff
@@ -203,6 +207,21 @@ impl win32::Host for EnvRef {
         }
     }
 
+    fn create_dir(&self, path: &WindowsPath) -> Result<(), u32> {
+        let path = windows_to_host_path(path);
+        std::fs::create_dir(path).map_err(|e| io_error_to_win32(&e))
+    }
+
+    fn remove_file(&self, path: &WindowsPath) -> Result<(), u32> {
+        let path = windows_to_host_path(path);
```
Comment (evmar) [inline cli/src/host.rs:216]: Not for this change, but it's probably worth starting to think about how we ought to model filesystem traversal.  For example can a rogue win32 app (or bug in the emulator) generate paths that make a `rm /r` into a `rm -rf /` outside?

**Review by evmar — COMMENTED** [review]
This all looks great except the one typo.

File: x86/src/ops/table.rs (hunk 1 above)
Comment (encounter) [reply x86/src/ops/table.rs]: Thanks, fixed

File: win32/src/winapi/kernel32/init.rs (hunk 2 above)
Comment (encounter) [reply win32/src/winapi/kernel32/init.rs:306]: A better heap implementation is going to be necessary soon, these compilers sometimes need more than 24mb of RAM :slightly_smiling_face:

File: cli/src/host.rs (hunk 5 above)
Comment (encounter) [reply cli/src/host.rs:216]: Thankfully there's no `rm -r` winapi (implemented, at least) that would make this an "easy" thing to happen. It'd have to recursively call DeleteFile, RemoveDirectory to make that happen.

**Comment by evmar** [comment]
Sorry I merged the other PR first and this conflicts, can you rebase?  🙏

**Comment by evmar** [comment]
I rebased it myself, nm!

**Pull request closed without merging**